subprojects/platform-play/src/main/java/org/gradle/play/internal/run/DefaultVersionedPlayRunAdapter.java

```diff
@@ -145,24 +142,41 @@ private void closeOldLoaders() throws IOException {
     public void buildSuccess() {
         reload.set(true);
         buildFailure = null;
-        makeReady(true);
+        pendingChanges.done();
     }
 
     @Override
     public void buildError(Throwable throwable) {
         buildFailure = throwable;
-        makeReady(true);
+        pendingChanges.done();
     }
 
     @Override
-    public void rebuildInProgress() {
-        makeReady(false);
+    public void expectPendingChanges() {
+        pendingChanges.more();
+
     }
 
-    private void makeReady(boolean isReady) {
-        synchronized (ready) {
-            ready.set(isReady);
-            ready.notifyAll();
+    private static class PendingChanges implements Serializable {
```

Comment:
I think it would be good if the client didn't need to track this stuff, and could just ask to be informed when both a) the build is complete and b) there's no further build in progress. Or perhaps just a simple "build complete" event that includes: success/failure, error, pending changes. Then the consumer can decide what to do for a completed build where pending changes are present.